worker/src/RTC/RtpPacket.cpp

```diff
@@ -600,12 +598,105 @@ namespace RTC
-				std::memset(extension->value + len, 0, currentLen - len);
+			MS_DEBUG_DEV(
+			  "set extension id: %" PRIu8 ", length:%" PRIu8 ", currentLen:%" PRIu8 " value:%s",
+			  id,
+			  len,
+			  currentLen,
+			  value.c_str());
+
+			// If len == currentLen, just memcpy.
+			if (len == currentLen)
+			{
+				std::memcpy(extension->value, value.c_str(), len);
+			}
+			// If len != currentLen, means we may need to shift extension or payload.
+			else
+			{
+				uint8_t* extensionStart    = reinterpret_cast<uint8_t*>(this->headerExtension) + 4;
+				uint8_t* extensionEnd      = extensionStart + GetHeaderExtensionLength();
+				uint8_t* ptr               = extensionStart;
+				size_t extensionsTotalSize = static_cast<size_t>(len + 1);
+				uint8_t* ptr1              = ptr;
+				// Clear current extension, we will rewrite after move all extension in right place.
+				std::memset((uint8_t*)extension, 0, currentLen + 1);
+				while (ptr < extensionEnd || ptr1 < extensionEnd)
+				{
+					// Means move has end. Fill the rest buffer to 0u.
+					if (ptr >= extensionEnd)
+					{
+						*ptr1 = 0u;
+						ptr1++;
+						continue;
+					}
+					const uint8_t tempExtensionId = (*ptr & 0xF0) >> 4;
+					const size_t tempExtensionLen = static_cast<size_t>(*ptr & 0x0F) + 1;
+
+					// tempExtensionId=15 in One-Byte extensions means "stop parsing here".
+					if (tempExtensionId == 15u)
+						break;
+					// Valid extension id.
+					if (tempExtensionId != 0u)
+					{
+						if (ptr1 < ptr)
+						{
+							std::memmove(ptr1, ptr, tempExtensionLen + 1);
+							// the extension has move from ptr to ptr1, store the One-Byte extension element in an
+							// array.
+							// `-1` because we have 14 elements total 0..13 and `id` is in the range 1..14.
+							this->oneByteExtensions[tempExtensionId - 1] =
+							  reinterpret_cast<OneByteExtension*>(ptr1);
+						}
+						extensionsTotalSize += tempExtensionLen + 1;
+						ptr += tempExtensionLen + 1;
+						ptr1 += tempExtensionLen + 1;
+
+						MS_DEBUG_DEV(
+						  "tempExtensionId: %" PRIu8 " tempExtensionLen:%zd, offset:%ld",
+						  tempExtensionId,
+						  tempExtensionLen,
+						  ptr - extensionStart);
+					}
+					// tempExtensionId=0 means alignment.
+					else
+					{
+						ptr++;
+					}
+				}
+
+				MS_DEBUG_DEV(
+				  "extensionsTotalSize: %zd headerLength:%zd",
+				  extensionsTotalSize,
+				  GetHeaderExtensionLength());
+
+				auto paddedExtensionsTotalSize =
+				  static_cast<size_t>(Utils::Byte::PadTo4Bytes(static_cast<uint16_t>(extensionsTotalSize)));
+				extensionsTotalSize = paddedExtensionsTotalSize;
+				int16_t shift = static_cast<int16_t>(extensionsTotalSize - GetHeaderExtensionLength());
+
+				MS_DEBUG_DEV("shift:%d paddedExtensionsTotalSize: %zd", shift, paddedExtensionsTotalSize);
 
-			// In One-Byte extensions value length 0 means 1.
-			extension->len = len - 1;
+				std::memmove(this->payload + shift, this->payload, this->payloadLength + this->payloadPadding);
```

Comment:
Is `payloadPadding` needed here? it's only valid to point the padding for parsed packets, but has no relevance otherwise.